strax/storage/s3.py

```diff
@@ -0,0 +1,217 @@
+        key_str = str(key)
+        bk = self.backend_key(key_str)
+
+        # Get list of all buckets / (run_id, lineage)
+        objects_list = self.s3client.list_buckets()
+
+        # If no buckets exist...
+        if 'Buckets' not in objects_list:
+            # No objects yet... so no metadata
+            if write:
+                return bk
+            else:
+                # If reading and no objects, then problem
+                raise strax.DataNotAvailable
+
+        # Loop over all buckets to look for our data
+        for key in objects_list['Buckets']:
+            if key['Name'] == key_str:
+                if write:
+                    if self._can_overwrite(key):
+                        return bk
+                    raise strax.DataExistsError(at=bk)
+                return bk
+
+        if write:
+            return bk
+        raise strax.DataNotAvailable
+
+    def backend_key(self, key_str):
+        return self.backends[0].__class__.__name__, key_str
+
+    def remove(self, key):
+        pass
+
+
+@export
+class S3Backend(strax.StorageBackend):
+    """Store data in S3-backend
+
+    Buckets are run/plugin keys and objects are chunks.
+    """
+
+    def __init__(self, **kwargs):
+        super().__init__()
+
+        #  Initialized connection to S3-protocol storage
+        self.s3client = boto3.client(**kwargs,
+                                     service_name='s3')
+        self.kwargs = kwargs
+
+    def get_metadata(self, backend_key):
+        result = self.s3client.get_object(Bucket=backend_key,
+                                          Key='metadata.json')
+
+        text = result["Body"].read().decode()
+        return json.loads(text)
+
+    def _read_chunk(self, backend_key, chunk_info, dtype, compressor):
+        with tempfile.SpooledTemporaryFile() as f:
+            self.s3client.download_fileobj(Bucket=backend_key,
+                                           Key=chunk_info['filename'],
+                                           Fileobj=f)
+            f.seek(0)  # Needed?
+            return strax.load_file(f, dtype=dtype, compressor=compressor)
+
+    def _saver(self, key, metadata, meta_only=False):
+        return S3Saver(key, metadata=metadata, meta_only=meta_only,
+                       **self.kwargs)
+
+
+@export
+class S3Saver(strax.Saver):
+    """Saves data to S3-compatible storage
+    """
+    json_options = dict(sort_keys=True, indent=4)
+
+    def __init__(self, key, metadata, meta_only,
+                 **kwargs):
+        super().__init__(metadata, meta_only)
+        self.s3client = boto3.client(**kwargs,
+                                     service_name='s3')
+
+        # Unique key specifying processing of a run
+        self.key = key
+
+        self.s3client.create_bucket(Bucket=self.key)  # Does nothing if exists
+
+    def _save_chunk(self, data, chunk_info):
+        filename = '%06d' % chunk_info['chunk_i']
+
+        with tempfile.SpooledTemporaryFile() as f:
```

Comment:
This file-writing code (open SpooledTemporaryfile, write to it, seek 0, upload to S3) is repeated three times, maybe consider adding a new method?

Reply:
I see what you mean, but hard to do.  One is download, one is upload JSON, one is upload chunk.  Therefore, the only line that is the same is the seek (which may not be needed...)

Reply:
There's another upload json at the end of close. But if you think avoiding duplication increases complexity too much, just leave it for now.

Reply:
Good point, within the Saver I can do that